src/SoftGlobalScope.jl

```diff
@@ -79,43 +79,44 @@ else
     recursively set to `true` for local scopes introduced by `for` etcetera.)
     NOTE: `_softscope`` may mutate the `globals` argument (if there are `local` declarations.)
     """
-    function _softscope(ex::Expr, globals, insertglobal::Bool=false)
+    function _softscope(ex::Expr, globals, locals, insertglobal::Bool=false)
         if isexpr(ex, :for) || isexpr(ex, :while)
-            return Expr(ex.head, ex.args[1], _softscope(ex.args[2], copy(globals), true))
+            return Expr(ex.head, ex.args[1], _softscope(ex.args[2], copy(globals), copy(locals), true))
         elseif isexpr(ex, :try)
-            try_clause = _softscope(ex.args[1], copy(globals), true)
-            catch_clause = _softscope(ex.args[3], ex.args[2] isa Symbol ? setdiff(globals, ex.args[2:2]) : copy(globals), true)
-            finally_clause = _softscope(ex.args[4], copy(globals), true)
+            try_clause = _softscope(ex.args[1], copy(globals), copy(locals), true)
+            catch_clause = _softscope(ex.args[3], copy(globals), ex.args[2] isa Symbol ? union!(locals, ex.args[2:2]) : copy(locals), true)
+            finally_clause = _softscope(ex.args[4], copy(globals), copy(locals), true)
             return Expr(:try, try_clause, ex.args[2], catch_clause, finally_clause)
         elseif isexpr(ex, :let)
-            letglobals = setdiff(globals, localvars(ex.args[1]))
-            return Expr(ex.head, _softscope(ex.args[1], globals, insertglobal),
-                                _softscope(ex.args[2], letglobals, true))
+            letlocals = union(locals, localvars(ex.args[1]))
+            return Expr(ex.head, ex.args[1], _softscope(ex.args[2], copy(globals), letlocals, true))
```

Comment:
Note that `args[1]` is problematic here.  e.g.
```jl
let x = (a += 1); end
```
needs a `global a` (assuming `:a in globals`)

Reply:
Yes, I know about this - it's mentioned in my original PR text above. The original code was also broken in this regard and so this is no less bad :smile: Plus this version doesn't have the other bug that was in the original (again, see above).